File: docs/JAMIA/comparisons/RPython/TM/26401116_median_inconsistent_types.R
# ID: SO#26401116
# Language: R
# Bug class: Value type
# Runtime consequence: Crash
# In study: Yes
# Inclusion rationale: Same int/double inconsistency from median() in data.table groupby. Not verified with R runtime — .R file written from SO code.
# Reproduction status: Fixed (data.table later enforces consistent return types)
# Type system catch: Grouped aggregation enforces consistent return type across groups

library(data.table)

DT <- data.table(
  V1 = c(2L, 3L, 1L, 1L, 1L, 0L),
  V2 = c(1L, 2L, 1L, 2L, 1L, 1L),
  V7 = factor(c(1, 2, 3, 2, 3, 3))
)

# median on integer columns: odd groups → integer, even groups → double → crash
DT[, lapply(.SD, median), by = V7]
# Error: Column 1 of result for group X is type 'integer' but expecting type 'double'.
# Column types must be consistent for each group.

File: docs/JAMIA/comparisons/RPython/TM/26401116_median_inconsistent_types.ts
/**
 * RPython SO#26401116 — Median returning an error when using data.table in R
 * Effect: Crash
 * Bug class: Int/double distinction
 *
 * Same pattern as 12125364. In R, median() returns different types depending on
 * group size. data.table requires consistent types across groups.
 *
 * In tidy-ts, number is number — no int/double distinction exists.
 * The type system still catches passing non-numeric columns to median.
 */
import { createDataFrame, stats as s } from "@tidy-ts/dataframe";

const DT = createDataFrame([
  { v1: 2, v2: 1, group: "1" },
  { v1: 3, v2: 2, group: "2" },
  { v1: 1, v2: 1, group: "3" },
  { v1: 1, v2: 2, group: "2" },
  { v1: 1, v2: 1, group: "3" },
  { v1: 0, v2: 1, group: "3" },
]);

const result = DT.groupBy("group").summarize({
  median_v1: (g) => s.median(g.extract("v1")),
  median_v2: (g) => s.median(g.extract("v2")),
});

// Passing the group column (string) to median is caught
// @ts-expect-error — string[] is not assignable to number[]
const wrong = s.median(DT.extract("group"));

result.print();
